plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs

```diff
@@ -0,0 +1,134 @@
+                                                       collectLiterals,
+                                                       getSrcSpan, getSrcText)
+import           Ide.Plugin.Retrie                    (handleMaybe,
+                                                       handleMaybeM, response)
+import           Ide.Types
+import           Language.LSP.Types
+import           Language.LSP.Types.Lens              (uri)
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionHandler
+    , pluginRules = collectLiteralsRule
+    }
+
+data CollectLiterals = CollectLiterals
+                     deriving (Show, Eq, Generic)
+
+instance Hashable CollectLiterals
+instance NFData CollectLiterals
+
+type instance RuleResult CollectLiterals = CollectLiteralsResult
+
+data CollectLiteralsResult = CLR {
+    literals      :: [Literal]
+    , formatTypes :: [FormatType]
+    } deriving (Generic)
+
+instance Show CollectLiteralsResult where
+    show _ = "<CollectLiteralResult>"
+
+instance NFData CollectLiteralsResult
+
+collectLiteralsRule :: Rules ()
+collectLiteralsRule = define $ \CollectLiterals nfp -> do
+    pm <- use GetParsedModule nfp
+    -- get the current extensions active and transform them into FormatTypes
+    let fmts = getFormatTypes <$> pm
+        -- collect all the literals for a file
+        lits = collectLiterals . pm_parsed_source <$> pm
+    pure ([], CLR <$> lits <*> fmts)
+    where
+        getFormatTypes = toFormatTypes . toList . extensionFlags . ms_hspp_opts . pm_mod_summary
+
+codeActionHandler :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionHandler state _ (CodeActionParams _ _ docId currRange _) = response $ do
+    nfp <- getNormalizedFilePath docId
+    CLR{..} <- requestLiterals state nfp
+        -- remove any invalid literals (see validTarget comment)
+    let litsInRange = filter inCurrentRange literals
+        -- generate alternateFormats and zip with the literal that generated the alternates
+        literalPairs = map (\lit -> (lit, alternateFormat formatTypes lit)) litsInRange
+        -- make a code action for every literal and its' alternates (then flatten the result)
+        actions = concatMap (\(lit, alts) -> map (mkCodeAction nfp lit) alts) literalPairs
+
+    logIO state $ "Literals: " <> show literals
+
+    pure $ List actions
+    where
+        inCurrentRange :: Literal -> Bool
+        inCurrentRange lit = let srcSpan = getSrcSpan lit
+                              in currRange `contains` srcSpan
+
+        mkCodeAction :: NormalizedFilePath -> Literal -> Text -> Command |? CodeAction
+        mkCodeAction nfp lit alt = InR CodeAction {
+            _title = "Convert " <> getSrcText lit <> " into " <> alt
+            -- what should this actually be?
+            , _kind = Just $ CodeActionUnknown "alternate.style"
```

Comment:
Great question. I'd love to be a bit more systematic about our code action kinds...

For this one, I think it should arguably go under `quickfix`. We use that not only for things that are "wrong" but also for simple alternative versions of things, e.g. all the import rewriting code actions. So I think something like `quickfix.literals.style` or something would be plausible.

Reply:
So just to clarify it would be `Just $ CodeActionUnknown "quickfix.literals.style"` or would it be `Just CodeActionQuickFix`?

Reply:
The former.